File: examples/fortran/example_05_hybrid_sigma_levels.f90
program example_05_hybrid_sigma_levels
  use cmip7_fortran_common
  implicit none

  integer, parameter :: nlev = 5
  character(len=1024) :: repo_root
  character(len=1024) :: output_dir
  character(len=2048) :: tables_path
  character(len=2048) :: input_path
  integer :: lat_id
  integer :: lon_id
  integer :: time_id
  integer :: lev_id
  integer :: ps_id
  integer :: var_id
  integer :: ierr
  integer :: nul_pos
  integer :: i
  integer :: j
  integer :: k
  integer :: t
  real :: cl(nlon, nlat, nlev, ntimes)
  real :: ps(nlon, nlat, ntimes)
  character(len=2048) :: filename
  double precision :: lat(nlat)
  double precision :: lon(nlon)
  double precision :: time(ntimes)
  double precision :: lat_bnds(2, nlat)
  double precision :: lon_bnds(2, nlon)
  double precision :: time_bnds(2, ntimes)
  double precision :: lev(nlev)
  double precision :: lev_bnds(nlev + 1)
  double precision :: a_coeff(nlev)
  double precision :: b_coeff(nlev)
  double precision :: a_bnds(nlev + 1)
  double precision :: b_bnds(nlev + 1)
  double precision :: p0

  call get_example_args(repo_root, output_dir)
  call prepare_cmor_paths(repo_root, output_dir, "example_05_input.json", &
       "mon", "r1", "f1", tables_path, input_path)

  ierr = cmor_setup(inpath=trim(tables_path), &
       netcdf_file_action=CMOR_REPLACE, &
       exit_control=CMOR_EXIT_ON_MAJOR)
  call check_status("cmor_setup", ierr)

  ierr = cmor_dataset_json(trim(input_path))
  call check_status("cmor_dataset_json", ierr)

  ierr = cmor_load_table("CMIP7_atmos.json")
  call check_id("cmor_load_table(CMIP7_atmos)", ierr)

  lat = (/ 10.0d0, 20.0d0, 30.0d0 /)
  lat_bnds(:, 1) = (/ 5.0d0, 15.0d0 /)
  lat_bnds(:, 2) = (/ 15.0d0, 25.0d0 /)
  lat_bnds(:, 3) = (/ 25.0d0, 35.0d0 /)

  lon = (/ 0.0d0, 90.0d0, 180.0d0, 270.0d0 /)
  lon_bnds(:, 1) = (/ -45.0d0, 45.0d0 /)
  lon_bnds(:, 2) = (/ 45.0d0, 135.0d0 /)
  lon_bnds(:, 3) = (/ 135.0d0, 225.0d0 /)
  lon_bnds(:, 4) = (/ 225.0d0, 315.0d0 /)

  time = (/ 15.5d0, 45.5d0 /)
  time_bnds(:, 1) = (/ 0.0d0, 31.0d0 /)
  time_bnds(:, 2) = (/ 31.0d0, 60.0d0 /)

  lon_id = cmor_axis(table="CMIP7_atmos.json", &
       table_entry="longitude", &
       units="degrees_east", &
       length=nlon, &
       coord_vals=lon, &
       cell_bounds=lon_bnds)
  call check_id("cmor_axis(longitude)", lon_id)

  lat_id = cmor_axis(table="CMIP7_atmos.json", &
       table_entry="latitude", &
       units="degrees_north", &
       length=nlat, &
       coord_vals=lat, &
       cell_bounds=lat_bnds)
  call check_id("cmor_axis(latitude)", lat_id)

  time_id = cmor_axis(table="CMIP7_atmos.json", &
       table_entry="time", &
       units="days since 1979-01-01", &
       length=ntimes, &
       coord_vals=time, &
       cell_bounds=time_bnds)
  call check_id("cmor_axis(time)", time_id)

  lev = (/ 0.92d0, 0.72d0, 0.50d0, 0.30d0, 0.10d0 /)
  lev_bnds = (/ 1.00d0, 0.83d0, 0.61d0, 0.40d0, 0.20d0, 0.00d0 /)
  lev_id = cmor_axis(table="CMIP7_atmos.json", &
       table_entry="standard_hybrid_sigma", &
       units="1", &
       length=nlev, &
       coord_vals=lev, &
       cell_bounds=lev_bnds)
  call check_id("cmor_axis(standard_hybrid_sigma)", lev_id)

  a_coeff = (/ 0.12d0, 0.22d0, 0.30d0, 0.20d0, 0.10d0 /)
  b_coeff = (/ 0.80d0, 0.50d0, 0.20d0, 0.10d0, 0.00d0 /)
  a_bnds = (/ 0.06d0, 0.18d0, 0.26d0, 0.25d0, 0.15d0, 0.00d0 /)
  b_bnds = (/ 0.94d0, 0.65d0, 0.35d0, 0.15d0, 0.05d0, 0.00d0 /)
  p0 = 100000.0d0

  ierr = cmor_zfactor(zaxis_id=lev_id, &
       zfactor_name="a", &
       axis_ids=(/ lev_id /), &
       zfactor_values=a_coeff, &
       zfactor_bounds=a_bnds)
  call check_id("cmor_zfactor(a)", ierr)

  ierr = cmor_zfactor(zaxis_id=lev_id, &
       zfactor_name="b", &
       axis_ids=(/ lev_id /), &
       zfactor_values=b_coeff, &
       zfactor_bounds=b_bnds)
  call check_id("cmor_zfactor(b)", ierr)

  ierr = cmor_zfactor(zaxis_id=lev_id, &
       zfactor_name="p0", &
       units="Pa", &
       zfactor_values=p0)
  call check_id("cmor_zfactor(p0)", ierr)

  ps_id = cmor_zfactor(zaxis_id=lev_id, &
       zfactor_name="ps", &
       axis_ids=(/ lon_id, lat_id, time_id /), &
       units="Pa")
  call check_id("cmor_zfactor(ps)", ps_id)

  do t = 1, ntimes
    do j = 1, nlat
      do i = 1, nlon
        ps(i, j, t) = 97000.0 + 400.0 * real(i - 1) &
             + 1600.0 * real(j - 1) + 100.0 * real(t - 1)
      enddo
    enddo
  enddo

  do t = 1, ntimes
    do k = 1, nlev
      do j = 1, nlat
        do i = 1, nlon
          cl(i, j, k, t) = 75.0 - 5.0 * real(k) - 1.2 * real(j - 1) &
               + 0.4 * real(i - 1) + 0.1 * real(t - 1)
        enddo
      enddo
    enddo
  enddo

  var_id = cmor_variable(table="CMIP7_atmos.json", &
       table_entry="cl_tavg-al-hxy-u", &
       units="%", &
       axis_ids=(/ lon_id, lat_id, lev_id, time_id /), &
       missing_value=missing_value)
  call check_id("cmor_variable(cl)", var_id)

  ierr = cmor_write(var_id, cl)
  call check_status("cmor_write(cl)", ierr)
  ierr = cmor_write(ps_id, ps, store_with=var_id)
  call check_status("cmor_write(ps)", ierr)

  filename = ""
  ierr = cmor_close(var_id, file_name=filename)
  call check_status("cmor_close(var)", ierr)
  nul_pos = index(filename, char(0))
  if (nul_pos > 0) filename(nul_pos:) = " "
  write(*, '(a)') trim(filename)

  ierr = cmor_close()
  call check_status("cmor_close", ierr)
end program example_05_hybrid_sigma_levels

File: examples/c/example_05_hybrid_sigma_levels.c
#include "cmip7_c_common.h"

#include <stdio.h>

#define NLEV 5

int main(int argc, char **argv) {
  const char *repo_root = argc > 1 ? argv[1] : ".";
  const char *output_dir = argc > 2 ? argv[2] : "examples/c/output";
  char input_path[CMIP7_PATH_MAX];
  char tables_path[CMIP7_PATH_MAX];
  char cell_measures[CMOR_MAX_STRING];
  char long_name[CMOR_MAX_STRING];
  char filename[CMOR_MAX_STRING];
  int file_action = CMOR_REPLACE;
  int exit_control = CMOR_EXIT_ON_MAJOR;
  int table_id, lon_id, lat_id, time_id, lev_id, ps_id, zfactor_id, var_id;
  int axes[4];
  int lev_axis[1];
  int ps_axes[3];
  int i, j, k, t;
  double lat[CMIP7_NLAT] = {10.0, 20.0, 30.0};
  double lat_bnds[CMIP7_NLAT + 1] = {5.0, 15.0, 25.0, 35.0};
  double lon[CMIP7_NLON] = {0.0, 90.0, 180.0, 270.0};
  double lon_bnds[CMIP7_NLON + 1] = {-45.0, 45.0, 135.0, 225.0, 315.0};
  double time[CMIP7_NTIMES] = {15.5, 45.5};
  double time_bnds[CMIP7_NTIMES + 1] = {0.0, 31.0, 60.0};
  double lev[NLEV] = {0.92, 0.72, 0.50, 0.30, 0.10};
  double lev_bnds[NLEV + 1] = {1.00, 0.83, 0.61, 0.40, 0.20, 0.00};
  double a_coeff[NLEV] = {0.12, 0.22, 0.30, 0.20, 0.10};
  double b_coeff[NLEV] = {0.80, 0.50, 0.20, 0.10, 0.00};
  double a_bnds[NLEV + 1] = {0.06, 0.18, 0.26, 0.25, 0.15, 0.00};
  double b_bnds[NLEV + 1] = {0.94, 0.65, 0.35, 0.15, 0.05, 0.00};
  double p0[1] = {100000.0};
  float cl[CMIP7_NTIMES * NLEV * CMIP7_NLAT * CMIP7_NLON];
  float ps[CMIP7_NTIMES * CMIP7_NLAT * CMIP7_NLON];
  float missing = CMIP7_MISSING_VALUE;

  snprintf(tables_path, sizeof(tables_path), "%s/cmip7-cmor-tables/tables",
           repo_root);
  cmip7_write_user_input_json(output_dir, "example_05_input.json", "mon", "r1",
                              "f1", input_path, sizeof(input_path));

  cmor_setup(tables_path, &file_action, NULL, &exit_control, NULL, NULL);
  cmor_dataset_json(input_path);
  cmor_load_table("CMIP7_atmos.json", &table_id);

  cmor_axis(&lon_id, "longitude", "degrees_east", CMIP7_NLON, lon, 'd',
            lon_bnds, 1, NULL);
  cmor_axis(&lat_id, "latitude", "degrees_north", CMIP7_NLAT, lat, 'd',
            lat_bnds, 1, NULL);
  cmor_axis(&time_id, "time", "days since 1979-01-01", CMIP7_NTIMES, time, 'd',
            time_bnds, 1, NULL);
  cmor_axis(&lev_id, "standard_hybrid_sigma", "1", NLEV, lev, 'd', lev_bnds, 1,
            NULL);

  lev_axis[0] = lev_id;
  cmor_zfactor(&zfactor_id, lev_id, "a", "", 1, lev_axis, 'd', a_coeff, a_bnds);
  cmor_zfactor(&zfactor_id, lev_id, "b", "", 1, lev_axis, 'd', b_coeff, b_bnds);
  cmor_zfactor(&zfactor_id, lev_id, "p0", "Pa", 0, NULL, 'd', p0, NULL);

  ps_axes[0] = time_id;
  ps_axes[1] = lat_id;
  ps_axes[2] = lon_id;
  cmor_zfactor(&ps_id, lev_id, "ps", "Pa", 3, ps_axes, 'f', NULL, NULL);

  for (t = 0; t < CMIP7_NTIMES; ++t) {
    for (j = 0; j < CMIP7_NLAT; ++j) {
      for (i = 0; i < CMIP7_NLON; ++i) {
        int idx = (t * CMIP7_NLAT + j) * CMIP7_NLON + i;
        ps[idx] = 97000.0f + 400.0f * (float)i + 1600.0f * (float)j +
                  100.0f * (float)t;
      }
    }
  }
  for (t = 0; t < CMIP7_NTIMES; ++t) {
    for (k = 0; k < NLEV; ++k) {
      for (j = 0; j < CMIP7_NLAT; ++j) {
        for (i = 0; i < CMIP7_NLON; ++i) {
          int idx = ((t * NLEV + k) * CMIP7_NLAT + j) * CMIP7_NLON + i;
          cl[idx] = 75.0f - 5.0f * (float)(k + 1) - 1.2f * (float)j +
                    0.4f * (float)i + 0.1f * (float)t;
        }
      }
    }
  }

  axes[0] = time_id;
  axes[1] = lev_id;
  axes[2] = lat_id;
  axes[3] = lon_id;
  cmor_variable(&var_id, "cl_tavg-al-hxy-u", "%", 4, axes, 'f', &missing, NULL,
                NULL, NULL, NULL, NULL);

  cmip7_get_cell_measures(tables_path, "atmos", "cl_tavg-al-hxy-u", "mon",
                          "glb", cell_measures, sizeof(cell_measures));
  cmor_set_variable_attribute(var_id, "cell_measures", 'c', cell_measures);
  if (cmip7_get_long_name_override(tables_path, "atmos", "cl_tavg-al-hxy-u",
                                   "mon", "glb", long_name,
                                   sizeof(long_name))) {
    cmor_set_variable_attribute(var_id, "long_name", 'c', long_name);
  }

  cmor_write(var_id, cl, 'f', NULL, CMIP7_NTIMES, NULL, NULL, NULL);
  cmor_write(ps_id, ps, 'f', NULL, CMIP7_NTIMES, NULL, NULL, &var_id);
  filename[0] = '\0';
  cmor_close_variable(var_id, filename, NULL);
  printf("%s\n", filename);
  cmor_close();
  return 0;
}
